Identify private torrents with no working trackers (GUI + WebUI)

an0n666 opened the pull request:
Closes #9849 

PR is mainly based on https://github.com/qbittorrent/qBittorrent/issues/9849#issuecomment-476034738

Thanks to @thalieht for allowing me to use his patch to make this PR.

I've fixed few minor issues with that patch and extended it to include WebUI and torrent filters.

Private torrents which do not have any working tracker will show in status "No Working Tracker" in red color. Such torrents can also be filtered by the Error torrent filter.

![Capture](https://user-images.githubusercontent.com/47396879/80602303-c1b36e00-8a50-11ea-8a64-6d0d4347f03f.PNG)

an0n666 commented on src/base/bittorrent/torrenthandleimpl.cpp:
```diff
@@ -173,6 +173,7 @@ TorrentHandleImpl::TorrentHandleImpl(Session *session, const lt::torrent_handle
     , m_ratioLimit(params.ratioLimit)
     , m_seedingTimeLimit(params.seedingTimeLimit)
     , m_hasSeedStatus(params.hasSeedStatus)
+    , m_hasWorkingTracker(true)
```
Using `m_hasWorkingTracker(hasWorkingTracker())` here causes updating/not contacted trackers to show as error on startup. That's why we're assuming torrent has a working tracker when added.
If anyone knows how to call that and not show error for updating/not contacted trackers then let me know.

FranciscoPombal commented:
:+1: 
Some comments:
1. I think the color should probably not be the same as a fatal error like missing files (see point 2).
2. I'm also not sure if private torrents with no working tracker should go to the errored category. It is my understanding that this category is reserved for errors with the torrent itself. "No working tracker" could be simply due to the tracker itself being down. Since it is impossible to determine whether or not the error is due to: the tracer being down, the torrent having been removed from it, or just a client-side connection error, I think this category should just be called "Warning".
3. The text in the status column should specify that this only applies to private torrents. E.g. "No working tracker (private torrent)" or "Private torrent with no working tracker". Yes, it is a little long (perhaps too long), but otherwise I can already see the issues users are going to post: "This torrent has no working trackers but does not go into the `No working trackers` status!!!", when this happens with public torrents.
4. This could probably be extended to public torrents under the following conditions: If DHT is off, public torrents are also put into the "Warning" category if they have no working tracker, but the status would instead say "no working tracker and DHT is off". Alternatively, the condition to move them to this state could be tightened to require all of DHT, LSD and PeX to be off, but I think DHT being down is far more important than the other 2, so that alone can be the condition.

an0n666 commented:
1. If a private torrent has no working tracker, I would consider that as a fatal error. 
2. I don’t think this deserves an additional filtering category. However if the devs insist then will add new filter. But bear in mind there’s already a lot of filters in the UI.
3. They could still create same issue even if we mentioned it. Anyways I will change the status message. I hope other devs will provide more input regarding this. 
4. I don’t really think public tracker users ever cared about trackers. Who uses public trackers without DHT anyway? Let’s keep that reserved for another PR. If some enthusiast public tracker user feels the need, they can extend the functionality.

FranciscoPombal commented:
@an0n666 
>     1. If a private torrent has no working tracker, I would consider that as a fatal error.

It can be a fatal error from the perspective of data transfer, but it is due to the unavailability of an external resource, not due to an error an the torrent handle level (like missing files and the like). The torrent itself is fine. But we should wait for more opinions on this.

>     2. I don’t think this deserves an additional filtering category. However if the devs insist then will add new filter. But bear in mind there’s already a lot of filters in the UI.

Yes, we should wait for more opinions.

>     3. They could still create same issue even if we mentioned it.

Disagree, I think we should make it obvious.

> Anyways I will change the status message. I hope other devs will provide more input regarding this.

:+1:
>     4. I don’t really think public tracker users ever cared about trackers. Who uses public trackers without DHT anyway? Let’s keep that reserved for another PR. If some enthusiast public tracker user feels the need, they can extend the functionality.

Agreed, it can be part of another PR. But in that case, my point (3) is even more important I think.

Ryrynz commented:
uTorrent simply colors the 'Seeding' bar red to display this, perhaps this is a good middle ground?

Piccirello commented:
Agreed with @FranciscoPombal on all points, but especially #3. The warning should be very specific about it pertaining only to private trackers.

On point #2, the existing "Stalled" status seems most appropriate to me. It definitely is not in the same class of errors to warrant the "Errored" status.

FranciscoPombal commented:
@Piccirello 

> the existing "Stalled" status seems most appropriate to me.

Disagree. The point of this PR is to be able to specifically isolate the private torrents that have no working tracker. This is slightly different, in general, from being "stalled". So, while I think the "Errored" category is not the most appropriate, I don't think "Stalled" is either. Plus a private torrent that loses tracker connection after getting the peers will still continue transferring data with the peers even after the tracker goes down, which means it is most definitely not stalled.

FranciscoPombal commented:
> Sometimes these things are temporary and we wouldn't want to give the impression something is very wrong.

:+1: This is a good summary of my point 2 above.

Piccirello commented:
>  private torrents that have no working tracker

I don't currently use private trackers so I guess I can't really speak to the use case. I'll stay out of this discussion. The only thing I caution against is adding a new status. The current list is already very long.

![Screen Shot 2020-04-30 at 12 43 39 PM](https://user-images.githubusercontent.com/8296030/80752526-79dc2780-8ae0-11ea-9992-589c5012f225.png)

FranciscoPombal commented:
@Piccirello 
> The only thing I caution against is adding a new status. The current list is already very long.

Hmm, you're not the first one to mention this concern.

I will say this: I vote for a brand new category, BUT: **if** it is decided that another category is too much, then **only in that case**, the most appropriate category is "Errored".

@glassez @sledgehammer999 @thalieht @NotTsunami thoughts?

an0n666 commented:
Just to add..this PR is currently not compatible with lt RC1_2. It’s using one deprecated function. I’ll fix try to fix it soon.

an0n666 commented:
I've added support for RC_1_2. Tested and it's working as expected.
Changed the status message to "No Working Tracker(Private Torrent)".

Now which colour should be used for such torrents?
And do we filter such torrents?
If yes, do we put them in any of the existing categories? (Error or Stalled) or do we add a new one?

Ryrynz commented:
'Stalled' would place it in with all the other stalled torrents so doesn't make it obvious, I still prefer a new category, call it whatever you like, 'Not working' makes sense to me.
As for the color, my vote is for the same red used for errored torrents.

an0n666 commented:
Stalled alone has 3 categories. That makes it difficult to add more filters. The list would be too long.

Ryrynz commented:
Why do we need a category just for just a sum of Stalled Upload and Stalled Download? I'd personally scrap 'Stalled' and replace it with 'Not working' or whatever, far more useful.. We can all do basic addition, I don't see it as useful.

FranciscoPombal commented:
FYI, I do understand the concern about "status bloat", but after investigating a bit myself, I don't think it is a problem just yet, even if this new category is added (I'm still not entirely dismissing the other viewpoint).

I resized my window so that the sidebar only has room for one more category (which would be this new one) before getting too big to fit every status without having to scroll. Here is what that looks like:

![screenshot](https://user-images.githubusercontent.com/17580742/80799297-698a8200-8b9e-11ea-9a26-ac4fc5815383.png)

The transfer list has room for 5 torrents, and without the details tab it would have room for 5 or 6 more. Do most people use their window this small, or even smaller, such that with the new category you'd have to scroll the sidebar to scroll all statuses?

thalieht commented:
I don't use private trackers either so i don't know what would be the best course of action for this.
However i believe there is a 99% chance that a new filter, only for private torrents and of this nature, will not be accepted in the STATUS section. Maybe, just _maybe_ in the TRACKERS section but since it's only for private torrents... i don't believe a new filter anywhere is acceptable.
I would suggest to take inspiration from other clients that have such conveniences for private tracker users.

FranciscoPombal commented:
@thalieht 
> I would suggest to take inspiration from other clients that have such conveniences for private tracker users.

Such as? (the specific conveniences, not the clients)

Ryrynz commented:
If it were to go into Stalled status and also into a new 'Not working' trackers section any chance we could get the trackers section to be shown in the WebUI?

thalieht commented:
>Such as? (the specific conveniences, not the clients)

Don't know anything about other clients. Just saw someone suggested a red progress bar from utorrent.

FranciscoPombal commented:
As I said above: IMO, if adding a new Status specifically for this out of the question, I vote for using the "Errored" category. This does imply that the scope of the "Errored" status will be broadened.

an0n666 commented:
I think private torrents that do not have any working tracker doesn't belong in stalled. As they could still be having active transfers due to transient tracker errors.
Also it shouldn't be in Errored status as it gives user a signal that something is wrong with the torrent itself.

I'm more inclined towards giving user a visual cue...Like the red progress bar in utorrent. Torrent state will remain as is.

glassez commented:
>Torrent state will remain as is.

:+1:
Torrent state should reflect the state of torrent itself.

FranciscoPombal commented:
@an0n666 
> I think private torrents that do not have any working tracker doesn't belong in stalled. As they could still be having active transfers due to transient tracker errors.
Also it shouldn't be in Errored status as it gives user a signal that something is wrong with the torrent itself.

Yep, agreed.

> I'm more inclined towards giving user a visual cue...Like the red progress bar in utorrent.

The problem with this is it does not allow the user to isolate the torrents with no working trackers. If I have 1000 torrents, I would like to have a view/filter which would only show the problematic torrents, not scroll through all the 1000 torrents looking for red progress bars.

One way to improve the "red progress bars" UX is to make sure they always sort either above 100% or below 0%. This way the user just has to sort torrents by progress and all of the problematic torrents will accumulate either at the top or bottom of the list.

@glassez 

> Torrent state will remain as is.

> Torrent state should reflect the state of torrent itself.

Sure, but does implementing a "Status" filter for private torrents with no working trackers imply defining a new state for that? Is it not possible to filter torrents based on other attributes other than state alone? In this case we just want to filter based on the following conditon: `is_private && no_working_trackers`.

LordNyriox commented:
@an0n666: 

> I'm more inclined towards giving user a visual cue...Like the red progress bar in utorrent. 

One problem with that suggestion is that some users may have their progress-bars red *by default*.  In particular, a modified "high contrast" Windows system theme may override the color value which the qBittorrent progress bar pulls from. 

**P.S.**  My own system-wide Windows theme has its highlight pattern set to "crimson on black" and its general text pattern set to "gray on black".  This results in crimson progress-bars in qBittorrent.

an0n666 commented:
I've re-worked on this PR and expanded it to include public torrents.

Torrents with no working tracker will show up as red in transfer list. That's the only visible change.
No changes to WebUI visuals to indicate a non working tracker. Maybe someone with more interest in the WebUI will make a PR for that.

Torrent status will reflect the actual status of the torrent, i.e - Seeding, [F] Seeding, Downloading, [F] Downloading etc.

No separate filter was added for such torrents. I'm still waiting for input from other devs as to if it's necessary, and if yes what should be the name.

![Annotation 2020-08-03 070115](https://user-images.githubusercontent.com/47396879/89139361-3b7f5a00-d560-11ea-9a62-43301314bdf4.png)

thalieht commented:
>expanded it to include public torrents.
>...
>Torrents with no working tracker will show up as red in transfer list in the GUI

Hey, half my torrents are trackerless. I don't want my list to be painted red for no reason.

thalieht commented on src/base/bittorrent/torrenthandle.h:
```diff
@@ -54,14 +54,20 @@ namespace BitTorrent
         Unknown = -1,
 
         ForcedDownloading,
+        ForcedDownloadingNWTracker,
```
Brevity is good but not at the expense of readability. If i read this outside the context of this PR i would have no idea what "Nw" means.

FranciscoPombal replied:
@thalieht pretty sure it means "Not working"

an0n666 replied:
I've updated them.

an0n666 commented:
> > expanded it to include public torrents.
> > ...
> > Torrents with no working tracker will show up as red in transfer list in the GUI
> 
> Hey, half my torrents are trackerless. I don't want my list to be painted red for no reason.

Will address that by ignoring trackerless torrents.

thalieht commented on src/base/bittorrent/torrenthandleimpl.cpp:
```diff
@@ -378,6 +379,22 @@ QVector<TrackerEntry> TorrentHandleImpl::trackers() const
+        return false;
+
+    for (const auto &tracker : trackers) {
+         const TrackerEntry &entry = tracker;
```
Redundant variable.

an0n666 replied:
Fixed

thalieht commented:
>Will address that by ignoring trackerless torrents.

That's still wrong IMO. Torrent's trackers may all be dead but there will always be DHT and PeX.

FranciscoPombal commented:
@an0n666 

This is a decent compromise solution, as long as trackerless torrents are not also painted red (i.e., trackerless torrents should be exempt from this logic).

I'm still convinced that a new filter would be the best solution, and that even without hiding empty filter/statuses, the list wouldn't be too bloated. But if others still disagree, we have to settle for this.

If this moves forward, I think you should also implement the WebUI part - traditionally, new "filter/states" additions have been implemented at the same time - no need to break feature parity now.

One ugly wart of this implementation is the addition of several `NWTracker`-suffixed repeated states. I doubt others will like this as well...

an0n666 commented:
> > Will address that by ignoring trackerless torrents.
> 
> That's still wrong IMO. Torrent's trackers may all be dead but there will always be DHT and PeX.

Are you suggesting to exclude public torrents when DHT is on or to exclude it altogether?

FranciscoPombal commented:
> > Will address that by ignoring trackerless torrents.
> 
> That's still wrong IMO. Torrent's trackers may all be dead but there will always be DHT and PeX.

Perhaps _public_ torrents with no working trackers could be colored orange/yellow (or not at all), and only _private_ torrents with no working trackers could be colored red?

thalieht commented:
>Are you suggesting to exclude public torrents when DHT is on or to exclude it altogether?

When DHT is on, good idea.

an0n666 commented:
> This is a decent compromise solution, as long as trackerless torrents are not also painted red (i.e., trackerless torrents should be exempt from this logic).
> 

Will be done. 

> I'm still convinced that a new filter would be the best solution, and that even without hiding empty filter/statuses, the list wouldn't be too bloated. But if others still disagree, we have to settle for this.
> 
Yes, a filter is handy to quickly find such torrents instead of the tracker error filter which picks up any torrent even with 1 failed tracker. 

> If this moves forward, I think you should also implement the WebUI part - traditionally, new "filter/states" additions have been implemented at the same time - no need to break feature parity now.
> 

I’ve updated the states in the WebUI. I just didn’t make any changes to color as I found the WebUI doesn’t actually implement a color by state function so I’ll have to implement it from scratch. 


> One ugly wart of this implementation is the addition of several `NWTracker`-suffixed repeated states. I doubt others will like this as well...
>

I’m sure there could be better ways to solve this issue. But no one seem to come up with one.

an0n666 commented:
> Perhaps _public_ torrents with no working trackers could be colored orange/yellow (or not at all), and only _private_ torrents with no working trackers could be colored red?
>

PR updated to use orange instead of red. Also excluded public torrents again as trackers are not the only point of failure. DHT turned on doesn't necessarily mean DHT is working. Also user may get peers through PeX/LSD or by manual addition. It's just not worth it to check public torrents.

![Annotation 2020-08-04 074435](https://user-images.githubusercontent.com/47396879/89244530-8c5a8580-d628-11ea-8bd5-74f0bc6de3f1.png)

ThatNerdyPikachu commented:
> Yes, a filter is handy to quickly find such torrents instead of the tracker error filter which picks up any torrent even with 1 failed tracker.

@an0n666 Will you be addressing this point in this PR, or should another one be opened? I'm happy to do it, if the latter.

xavier2k6 commented:
What's the status of this PR now?, since it seems @an0n666's account has been deleted!

xavier2k6 commented:
@an0n666 - you're back!, cool - is this PR abandoned or still a WIP?

an0n666 commented:
> @an0n666 - you're back!, cool - is this PR abandoned or still a WIP?

The former.
Currently, I don't have anything in mind to make a cleaner implementation.
Maybe someone with more experience with qBt internals will take it over at some point.

sakkamade commented:
Please do not mix private _torrents_ with torrents with private/semi-private _trackers_.

I very much wonder that the issue 9849 (which mentioned above) is regarding former.
Why should one wish to highlight the _private torrent_ when they may only download and not upload it, thus should probably delete it right upon downloading?

sakkamade commented:
> > Will address that by ignoring trackerless torrents.
> 
> 
 > That's still wrong IMO. Torrent's trackers may all be dead but there will always be DHT and PeX.



>  > Are you suggesting to exclude public torrents when DHT is on or to exclude it altogether?
>
> When DHT is on, good idea. I'm not sure if PeX can work without DHT though, maybe both? Or just DHT if you know PeX can't work without it.

The whole point of red-, or orange-coloured torrent is to show that it has no working tracker, it is not a question of whether user (or torrent) has DHT enabled or not. In other words, this is not the question of whether user can continue to seed it or not.

In most cases, if torrent has no working tracker then the torrent was updated or removed, wherever user  had downloaded it from.
If updated, user will download new torrent, and replace it with old one.
If removed, user may continue to seed it (unless torrent is private), since there is also a DHT, to delete it, etc.

I should think that the majority of users who wishing this feature is wishing it for the first case.

This feature is useful for public torrents just as much as for private.
